crates/engine/local/src/service.rs

```diff
@@ -0,0 +1,322 @@
+    payload_builder: PayloadBuilderHandle<N::Engine>,
+    /// The attributes for the payload builder
+    payload_attributes: <N::Engine as PayloadTypes>::PayloadAttributes,
+    /// A handle to the persistence layer
+    persistence_handle: PersistenceHandle,
+    /// The hash of the current head
+    head: B256,
+    /// The mining mode for the engine
+    mode: MiningMode,
+    /// Payload request task
+    payload_request_state: RequestState,
+}
+
+impl<N: EngineNodeTypes> std::fmt::Debug for LocalEngineService<N> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("LocalEngineService")
+            .field("payload_builder", &self.payload_builder)
+            .field("payload_attributes", &self.payload_attributes)
+            .field("persistence_handle", &self.persistence_handle)
+            .field("head", &self.head)
+            .field("mode", &self.mode)
+            .field("payload_request_state", &"...")
+            .finish()
+    }
+}
+
+/// The internal state of the [`LocalEngineService`]'s polling task
+#[derive(Default)]
+struct RequestState {
+    ready: bool,
+    task: Option<BoxFuture<'static, eyre::Result<B256>>>,
+}
+
+impl<N> LocalEngineService<N>
+where
+    N: EngineNodeTypes,
+{
+    /// Constructor for [`LocalEngineService`].
+    pub fn new(
+        payload_builder: PayloadBuilderHandle<N::Engine>,
+        payload_attributes: <N::Engine as PayloadTypes>::PayloadAttributes,
+        provider: ProviderFactory<N>,
+        pruner: Pruner<N::DB, ProviderFactory<N>>,
+        head: B256,
+        mode: MiningMode,
+    ) -> Self {
+        let persistence_handle = PersistenceHandle::spawn_service(provider, pruner);
+
+        Self {
+            payload_builder,
+            payload_attributes,
+            persistence_handle,
+            head,
+            mode,
+            payload_request_state: RequestState::default(),
+        }
+    }
+
+    /// Spawn a [`LocalEngineService`] on a tokio green thread.
+    pub fn spawn_new(
+        payload_builder: PayloadBuilderHandle<N::Engine>,
+        payload_attributes: <N::Engine as PayloadTypes>::PayloadAttributes,
+        provider: ProviderFactory<N>,
+        pruner: Pruner<N::DB, ProviderFactory<N>>,
+        head: B256,
+        mode: MiningMode,
+    ) {
+        let engine = Self::new(payload_builder, payload_attributes, provider, pruner, head, mode);
+
+        // Spawn the engine
+        tokio::spawn(engine);
+    }
+}
+
+/// Run the [`LocalEngineService`] as a Future. The service will poll the payload builder
+/// with two varying modes, [`MiningMode::Instant`] or [`MiningMode::Interval`]
+/// which will respectively either execute the block as soon as it finds a
+/// transaction in the pool or build the block based on an interval.
+impl<N> Future for LocalEngineService<N>
+where
+    N: EngineNodeTypes,
+{
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
```

Comment:
if it's easier we can also do this as an async fn function over tokio select, which would make sense here because we do a lot of sequential processing